stoppable_workers.go

```diff
@@ -0,0 +1,64 @@
+package utils
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrStoppableWorkersAlreadyStopped is returned by Add when the StoppableWorkers
+// instance has already been stopped.
+var ErrStoppableWorkersAlreadyStopped = errors.New("cannot add worker: already stopped")
+
+// StoppableWorkers is a collection of goroutines that can be stopped at a
+// later time.
+type StoppableWorkers struct {
+	mu         sync.RWMutex
+	ctx        context.Context
+	cancelFunc func()
+
+	workers sync.WaitGroup
+}
+
+// NewStoppableWorkers creates a new StoppableWorkers instance. The instance's
+// context will be derived from passed in context.
+func NewStoppableWorkers(ctx context.Context) *StoppableWorkers {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	return &StoppableWorkers{ctx: ctx, cancelFunc: cancelFunc}
+}
+
+// Add starts up a goroutine for the passed-in function. Workers:
+//
+//   - MUST respond appropriately to errors on the context parameter.
+//   - MUST NOT add more workers to the `StoppableWorkers` group to which
+//     they belong.
+//
+// Any `panic`s from workers will be `recover`ed and logged.
+func (sw *StoppableWorkers) Add(worker func(context.Context)) error {
```

Comment:
I remember Alan had to make it take a variardic input to determine the additional workers at runtime; in one commit, you can find them in the StoppableWorkers' file commit history.

Reply:
I think he made it variardic so that you could pass in many functions to use the same context,  but I guess this works the same now that I think about it. It'll just have to be called in a loop.